_posts/2020-03-29-all-you-need-is-lambda-1-booleans.md

```diff
@@ -0,0 +1,260 @@
+
+> Aside: Due to the lambda calculus’ terseness, I will be making free use of several notational conveniences:
+>
+> 1. writing `λ x y . z` as an abbreviation of `λ x . λ y . z`.
+> 2. writing `?` to stand for bits we don’t know yet, as tho we had an environment supporting holes.
+> 3. writing definitions as tho we had a metalanguage.
+> 4. referencing definitions elsewhere as tho we had globals.
+> 5. writing type signatures as tho we had a type system, and even a typechecker, with as much polymorphism and inference as is convenient at any particular moment.
+> 6. using syntactic recursion as tho it existed.
+> 7. using general recursion as tho it made sense.
+> 8. ignoring application order, normalization, reduction, substitution, values, references, allocation, copying, space, time, entropy, and any and all other such details whenever I feel like it.
+>
+> Further, I will by convention name types in `TitleCase` and both term and (local) type variables in `camelCase`.
+>
+> I will try to avoid pulling rabbits from hats too wantonly, but for now, I’ll ask you to suspend disbelief; I hope to revisit and justify some of these in later posts.
+
+
+## λ is blind
+
+Of course, it’s not _just_ booleans we’re after; `true` and `false` aren’t much use without `and`, `or`, `not`, `if`, and all the rest. To be useful, our representation of booleans should therefore suffice to define these, as well. But how do you define `if` without using `if`? In a lazy language like Haskell, we might define `if` as a function something like so:
+
+```haskell
+if_ :: Bool -> a -> a -> a
+if_ cond  then_ else_ = if cond then then_ else else_
+```
+
+In a strict language like JavaScript, we’d instead take functions for the alternatives:
+
+```javascript
+function if_(cond, then_, else_) {
+  if (cond) {
+    then_();
+  } else {
+    else_();
+  }
+}
+```
+
+Both these definitions use the language’s native booleans and `if` syntax (a tactic for implementing embedded DSLs known as “meta-circularity”), and thus aren’t viable in the lambda calculus. However, they do give us a hint: in both cases we have a function taking a condition, consequence, and alternative, and using the first to select one of the latter two. In the lambda calculus, we might start by writing:
+
+```
+if = λ cond then else . ?
+```
+
+(Note: there aren’t any keywords in the lambda calculus, so there’s nothing stopping me from naming variables things like `if`, a fact which I will take free advantage of.)
+
+We’ve introduced a definition for `if`, as a function of three parameters; now what do we do with them? The lambda calculus’ stark palette makes it easy to enumerate _all_ the things we can do with a variable:
+
+1. Ignore it, or shadow it with another lambda binding the same name.
```

Comment:
```suggestion
1. Ignore it, or shadow it with another lambda binding of the same name.
```

Reply:
Binding is used as a verb here, not a noun.

Reply:
Fair!